extensions/amp-analytics/0.1/variables.js

```diff
@@ -0,0 +1,263 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {cryptoFor} from '../../../src/crypto';
+import {dev, user} from '../../../src/log';
+import {fromClass} from '../../../src/service';
+import {isArray, isFiniteNumber, map} from '../../../src/types';
+
+/** @const {String} */
+const TAG = 'Analytics.Variables';
+
+/** @const {RegExp} */
+const VARIABLE_ARGS_REGEXP = /([^(]*)(\([^)]*\))?/;
+
+/** @typedef {function(...?):(?|!Promise<?>)} */
+let VariableFilterDef;
+
+/**
+ * @param {!string} str
+ * @param {Number} s
+ * @param {Number} l
+ * @return {string}
+ */
+function substrFilter(str, s, l) {
+  const start = Number(s);
+  let length = str.length;
+  user().assert(isFiniteNumber(start),
+      'Start index ' + start + 'in substr filter should be a number');
+  if (l) {
+    length = Number(l);
+    user().assert(isFiniteNumber(length),
+        'Length ' + length + ' in substr filter should be a number');
+  }
+
+  return str.substr(start, length);
+}
+
+/**
+ * @param {*} value
+ * @param {*} defaultValue
+ * @return {*}
+ */
+function defaultFilter(value, defaultValue) {
+  return value || user().assertString(defaultValue);
+}
+
+
+export class VariableService {
+  /**
+   * @param {!Window} window
+   */
+  constructor(window) {
+
+    /** @private {!Window} */
+    this.win_ = window;
+
+    /** @private {!Object<string, VariableFilterDef>} */
+    this.filters_ = map();
+
+    this.register_('default', defaultFilter);
+    this.register_('substr', substrFilter);
+    this.register_('trim', value => user().assertString(value).trim());
+    this.register_('json', value => JSON.stringify(value));
+    this.register_('toLowerCase', value =>
+        user().assertString(value).toLowerCase());
+    this.register_('toUpperCase', value =>
+        user().assertString(value).toUpperCase());
+    this.register_('not', value => String(!value));
```

Comment:
Yes, I think it results in consistency. Otherwise falsey values give value: `true` whereas truthy values give a value: `` (empty string)